lib/Support/Progress.cpp

```diff
@@ -0,0 +1,255 @@
+/// \file Progress.cpp
+/// \brief
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <chrono>
+
+#include "llvm/Support/Progress.h"
+
+#include "revng/Support/Debug.h"
+
+class TraceProgressListener : public llvm::ProgressListener {
+private:
+  std::vector<std::optional<std::string>> LastStepName;
+  std::error_code EC;
+  llvm::raw_fd_ostream Output;
+
+public:
+  TraceProgressListener(llvm::StringRef OutputPath)
+    : Output(OutputPath, EC) {
+    revng_assert(!EC);
+    Output << "[\n";
+  }
+
+  ~TraceProgressListener() override {
+    Output << "{}]\n";
+  }
+
+public:
+  void handleNewTask(const llvm::Task *T) override {
+    LastStepName.emplace_back();
+    emitEvent(T->name(), "task", "B");
+  }
+
+  void handleTaskCompleted(const llvm::Task *T) override {
+    endStep();
+    LastStepName.pop_back();
+    emitEvent(T->name(), "task", "E");
+  }
+
+  void handleTaskAdvancement(const llvm::Task *T,
+                             llvm::StringRef NewStepName) override {
+    endStep();
+    emitEvent(NewStepName, "task", "B");
+    LastStepName.back() = NewStepName;
+  }
```

Comment:
These three methods are advertised  in the sister llvm MR to be callable by multiple threads at the same time. 

If this happens, there are 3 issues:
- `emitEvent` does not guard against concurrent access to `Output` so you could end up printing garbled output;
- `LastStepName` is not guarded against concurrent access, so it could end up in invalid state.
- `endStep` is also not guarded against concurrent access, so it's not safe.

I think you should put a `llvm::sys::SmartRWMutex` in `TraceProgressListener` and use it to make this API safe for concurrent use.

Reply:
BTW, all other `ProgressListeners` in this file have the same flaw.

Reply:
Yep, I'm aware. This is still WIP.

* `LastStepName` has to go. I changed the API on LLVM side so you no longer need it.
* My plan is to lock all writes to the output for `PlainProgressListener` and `TerminalBarsProgressListener` (once every 50ms). While I thinking of buffering output in thread-specific buffers for `TraceProgressListener`. The ratio is that the `PlainProgressListener` might be useful for debugging a crash, so delaying output is not helpful. `TerminalBarsProgressListener` is already amortized. `TraceProgressListener` is more for performance profiling than debugging stuff, so I think buffering the output is fine.

Reply:
It works for this listener, but in general all listeners that have internal state need to be built to be safe to be called concurrently.

Reply:
I put a lock in front of `TraceProgressListener` and introduced a `bool` to state that the other listeners are interested in the main thread only, which bypasses the threading problem entirely.